src/components/tl/nccl/tl_nccl_coll.c

```diff
@@ -64,6 +69,82 @@ static inline ucc_status_t ucc_nccl_check_dt_supported(ucc_datatype_t dt1,
     return UCC_OK;
 }
 
+ucc_tl_nccl_task_t * ucc_tl_nccl_init_task(ucc_base_coll_args_t *coll_args,
+                                           ucc_base_team_t *team)
+{
+    ucc_tl_nccl_context_t *nccl_ctx  = ucc_derived_of(team->context,
+                                                      ucc_tl_nccl_context_t);
+    ucc_tl_nccl_task_t    *task;
+    ucc_status_t           status;
+
+    task = ucc_mpool_get(&nccl_ctx->req_mp);
+    ucc_coll_task_init(&task->super, coll_args, team);
+    UCC_TL_NCCL_PROFILE_REQUEST_NEW(task, "tl_nccl_task", 0);
+    task->super.finalize           = ucc_tl_nccl_coll_finalize;
+    task->super.triggered_post     = ucc_tl_nccl_triggered_post;
+    task->completed                = NULL;
+    task->allgatherv_bcopy.scratch = NULL;
+    if (nccl_ctx->cfg.sync_type == UCC_TL_NCCL_COMPLETION_SYNC_TYPE_EVENT) {
+        status = ucc_mc_ee_create_event(&task->completed, UCC_EE_CUDA_STREAM);
+        if (ucc_unlikely(status != UCC_OK)) {
+            ucc_mpool_put(task);
+            return NULL;
+        }
+    }
+    return task;
+}
+
+void ucc_tl_nccl_free_task(ucc_tl_nccl_task_t *task)
+{
+    UCC_TL_NCCL_PROFILE_REQUEST_FREE(task);
+    ucc_mpool_put(task);
+}
+
+ucc_status_t ucc_tl_nccl_triggered_post(ucc_ee_h ee, ucc_ev_t *ev,
+                                        ucc_coll_task_t *coll_task)
+{
+    ucc_tl_nccl_task_t *task  = ucc_derived_of(coll_task, ucc_tl_nccl_task_t);
+    ucc_status_t status;
+    ucc_ev_t *post_event;
+
+    ucc_assert(ee->ee_type == UCC_EE_CUDA_STREAM);
+    coll_task->ee = ee;
+    tl_info(UCC_TASK_LIB(task), "triggered post. task:%p", coll_task);
+
+    status = coll_task->post(coll_task);
+    if (ucc_likely(status == UCC_OK)) {
+        /* TODO: mpool */
+        post_event = ucc_malloc(sizeof(ucc_ev_t), "event");
+        if (ucc_unlikely(post_event == NULL)) {
+            tl_error(UCC_TASK_LIB(task), "failed to allocate memory for event");
+            return UCC_ERR_NO_MEMORY;
+        }
+
+        post_event->ev_type = UCC_EVENT_COLLECTIVE_POST;
+        post_event->ev_context_size = 0;
+        post_event->req = &coll_task->super;
+        ucc_ee_set_event_internal(coll_task->ee, post_event,
+                                  &coll_task->ee->event_out_queue);
+    }
+    return status;
+}
+
+ucc_status_t ucc_tl_nccl_coll_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_nccl_task_t *task  = ucc_derived_of(coll_task, ucc_tl_nccl_task_t);
+    ucc_status_t       status = UCC_OK ;
+
+    tl_info(UCC_TASK_LIB(task), "finalizing coll task %p", task);
+    if (task->completed) {
+        ucc_mc_ee_destroy_event(task->completed, UCC_EE_CUDA_STREAM);
+    }
+    if (task->allgatherv_bcopy.scratch) {
```

Comment:
same here: if we add another alg under union then it is not defined here, better put into bcopy finalize i think.